### DISCUSSION TITLE
Removing Version Suffix from all jars

### DISCUSSION THREAD

**Discussion by Lexian-droid** [EssentialsX/Essentials#5187] [Ideas and Feature Suggestions] [answered]
Hello lads, I have a quick suggestion which would make automation for updating Essentials much better.
I was thinking, every new release of Essentials has it's version suffixed on the filename, this makes it more difficult to automate the process, because in order to download the new update, you now have to change it to match the new version.

I was thinking, it would be great if there was a change with how the files are listed, the directory structure, renamed all jar files to exclude any version names, and just leaving "Version" inside of the meta data.

This is similar approach to plugins like Geyser, or Spigot.
It would improve usability of these plugins a lot, as there are a lot of Jar files and there are a lot of updates.

It would simply make life much easier, because instead of using the exact filename, we can do something like, https://github.com/EssentialsX/Essentials/releases/download/latest/EssentialsX.jar which will automatically, if the configuration is accepted; should automatically give the latest version, and it would just greatly improve how we use this plugin every day in our servers.

Hope this finds some interest, thank you for reading, and have a good day.

**Comment by Lexian-droid** [comment]
Correction: This is a similar approach to Geyser and *Floodgate*
Which both just have one name when you download it, thus making it very easy to update them automatically.

**Comment by mdcfe** [comment, marked as answer]
The version name is an important piece of information when you're updating plugin jars, as it tells you what versions you have at a glance (without starting up the server or opening the `plugin.yml` to check). We will not be removing these from our published jar file names.

It is bad practice to automate any plugin updates in the first place - you should check what's changed in recent dev builds or read the full changelog for stable builds, then take backups of EssentialsX data and apply any config changes as required to avoid losing any data. Automatic updates are a recipe for disaster and not something we will facilitate.

When updating plugins manually, it is incredibly trivial to move files from a shell. Once you're confident that the new versions are suitable and your data is safely backed up, you can move or rename your existing jars with a wildcard (don't delete old jars until you've tested the new builds - you might need to roll back). This makes it largely redundant to remove the version string from the file name.

**Reply by Lexian-droid** [reply]
I believe "leaving the old jars" is impossible, this will cause a version error in Spigot.

I understand what you mean, however I personally never needed to look at the version string, because if I'm going through the effort of downloading all these new builds, I already know what versions are uploaded to the server. However, I understand the practicality, and while I personally disagree with your assessment, I suppose it is how the cookie crumbles.

I do want to add, I feel as though essentials is either, too invasive with updates, or it gets updates too often. When there is 1 update, you have to upload many jar files, in order to prevent the update check. In addition to that - It should be detecting updates from the stable branch, not the beta branch. as well as that; the plugin writes an update monologue all to say you're 1 version out of date, Which fills up the screen of the users with essentials update check permission, and the console.

I also noticed I was downloading beta branches by mistake, as it was the default selected tab on the website. I don't know if that's why they are updating so much, but when I switched to the stable branch, I got errors that read "This version isn't supported" so if I had to guess, you guys are working on EssentialsX 1.19.3?  Regardless, the only reason I made the suggestion was just to have an easier swap-in-swap-out, similar to geyser, and you don't really see people complaining about that.

I appreciate your response though, thank you.

**Comment by Bobcat00** [comment]
I'm OK with the version suffix, but I wish all the .jar files could be downloaded together in a ZIP file. That way you're guaranteed to get a matching set and you could simply extract the .jar files you want to use. It's also a convenient way to keep the matching files together as a backup.

**Reply by Evidentsinger14** [reply]
this used to be the case, however people were installing ALL jars (AntiBuild, XMPP, and Protect aren't used nearly as much), and it caused an issue.

**Reply by mdcfe** [reply]
To add to this, the jars for every module on the download page will always be the same version as one another.

**Comment by Chew** [comment]
If you use Paper, you can just toss all your plugin JARs into the `update` folder. Paper is smart in modern versions, and it will overwrite the JARs based on the plugin name, not JAR/file name, further rendering this suggestion as unnecessary.

**Reply by Lexian-droid** [reply]
This is a feature? could you link me some information, as I cannot find any information in the documentation or anything. Much appreciated!

**Reply by Bobcat00** [reply]
It's Paper! There's no documentation - You have to read code and figure it out yourself!
https://github.com/PaperMC/Paper/blob/1143b636634183fe9ee6d8cf9709afd4b3239de4/patches/api/0377-Update-Folder-Uses-Plugin-Name.patch

**Reply by Lexian-droid** [reply]
So, the trouble is, I know nothing about java... While I do know syntax of average code, from what I see is kind of just, a bunch of checks and functions... then from what it appears, it will change the name of the file, relocate it, and delete something. It returns the new file name, ... and restarts the plugin loader?

Could you maybe inform me as to how i would actually get this to work? I am checking my paper installation, and I don't really see anything like "update" folder anywhere.

**Reply by mdcfe** [reply]
This isn't Paper support. You can use the `plugins/update` folder if you want, but that's beyond the scope of this feature request.

**Reply by Bobcat00** [reply]
The folder is `plugins/update`, all lower case. If you don't have it, just create it. (Also note this is **not** the `Updater` folder which may also be present.)

So you can just stick plugin .jar files that need to be updated in the `update` folder while your server is running. The next time you restart your server, it will automatically move those files to the real `plugins` folder for plugins with matching names.

This partial name matching for Paper is news to me, so I don't really know how it works. It looks like it's grabbing the plugin name from each plugin's plugin.yml file and using that to make the comparison, then renaming the file after it's moved so the filename matches the name in plugin.yml. WAIT! But that would mean EssentialsX.jar would be renamed to Essentials.jar, _so I have no idea what's going on_.

**Reply by Lexian-droid** [reply]
> The folder is `plugins/update`, all lower case. If you don't have it, just create it. (Also note this is **not** the `Updater` folder which may also be present.)
> 
> So you can just stick plugin .jar files that need to be updated in the `update` folder while your server is running. The next time you restart your server, it will automatically move those files to the real `plugins` folder for plugins with matching names.
> 
> This partial name matching for Paper is news to me, so I don't really know how it works. It looks like it's grabbing the plugin name from each plugin's plugin.yml file and using that to make the comparison, then renaming the file after it's moved so the filename matches the name in plugin.yml. WAIT! But that would mean EssentialsX.jar would be renamed to Essentials.jar, _so I have no idea what's going on_.

Very helpful, thank you, I will look into this! Thank you very much.



> This isn't Paper support. You can use the `plugins/update` folder if you want, but that's beyond the scope of this feature request.

You're right, quite right, I wasn't asking for support here regarding paper either, someone simply mentioned the existence of an update folder, and I simply wanted to request more information. and some people were kind enough to give some additional information. Please, us developers should support each other, don't bash a user for simply asking about information for something that is actually pretty closely related.

**Reply by Chew** [reply]
> This is a feature? could you link me some information, as I cannot find any information in the documentation or anything. Much appreciated!

https://docs.papermc.io/paper/updating#step-2-update-plugins
